modules/ndu/src/backend/ndu-engine.ts

```diff
@@ -5,84 +5,49 @@ import _ from 'lodash'
 
 const debug = DEBUG('ndu').sub('processing')
 
-const ActionTypes = <const>[
-  'faq_trigger_outside_topic',
-  'faq_trigger_inside_topic',
-  'wf_trigger_inside_topic',
-  'wf_trigger_outside_topic',
-  'wf_trigger_inside_wf',
-  'faq_trigger_inside_wf',
-  'node_trigger_inside_wf'
-]
-
-type ActionType = typeof ActionTypes[number]
-type ActionPredictions = { [key in ActionType]: number }
-
-const stringToVec = (choices: string[], value: string) => {
-  const arr = Array(choices.length).fill(0)
-  const idx = choices.indexOf(value)
-  if (idx >= 0) {
-    arr[idx] = 1
-  }
-  return arr
-}
-
-const newFeature = (): Features => ({
-  conf_faq_trigger_inside_topic: 0,
-  conf_faq_trigger_outside_topic: 0,
-  conf_faq_trigger_parameter: 0,
-  conf_node_trigger_inside_wf: 0,
-  conf_wf_trigger_inside_topic: 0,
-  conf_wf_trigger_inside_wf: 0,
-  conf_wf_trigger_outside_topic: 0,
-  current_highest_ranking_trigger_id: '',
-  current_node_id: '',
-  current_workflow_id: '',
-  last_turn_action_name: '',
-  last_turn_same_highest_ranking_trigger_id: false,
-  last_turn_same_node: false,
-  last_turn_since: 0
-})
-
-const dataset: [Features, ActionType][] = BASE_DATA.map(([feat, label]) => [
-  Object.assign(newFeature(), (feat as any) as Features),
-  <ActionType>label
-])
-
-const WfIdToTopic = (wfId: string): string | undefined => {
-  if (wfId === 'n/a' || wfId.startsWith('skills/')) {
-    return
-  }
-
-  return wfId.split('/')[0]
-}
-
 export const DEFAULT_MIN_CONFIDENCE = 0.1
 
 export class UnderstandingEngine {
   private _allTopicIds: Set<string> = new Set()
   private _allNodeIds: Set<string> = new Set()
   private _allWfIds: Set<string> = new Set()
 
-  private _allTriggers: Map<string, sdk.NDU.Trigger[]> = new Map()
+  private _allTriggers?: sdk.NDU.Trigger[]
   private _minConfidence: number
 
-  trainer: sdk.MLToolkit.SVM.Trainer
-  predictor: sdk.MLToolkit.SVM.Predictor
-
-  constructor(private bp: typeof sdk, private _dialogConditions: sdk.Condition[], config: Config) {
-    this.trainer = new this.bp.MLToolkit.SVM.Trainer()
+  private predictor: sdk.MLToolkit.SVM.Predictor
+  private trainer: Trainer
+  private loadedModelHash: string
+
+  constructor(
+    private bp: typeof sdk,
+    private botId: string,
+    private _dialogConditions: sdk.Condition[],
+    config: Config
+  ) {
+    this.trainer = new Trainer(bp, botId)
     this._minConfidence = config.minimumConfidence ?? DEFAULT_MIN_CONFIDENCE
   }
 
+  async loadModel(model: ModelService.Model) {
+    if (this.loadedModelHash !== undefined && this.loadedModelHash === model.hash) {
+      return
+    }
+
+    this.predictor = new this.bp.MLToolkit.SVM.Predictor(model.data)
+    this.loadedModelHash = model.hash
+  }
+
   featToVec(features: Features): number[] {
     const triggerId = stringToVec(
-      _.flatten([...this._allTriggers.values()].map(x => x.map(this.getTriggerId))),
+      _.flatten(this._allTriggers.map(x => getTriggerId(x))),
```

Comment:
what was the reason why it was copied in an array at first ?

Reply:
Because allTriggers was used to store all bots, now the engine itself is scoped to the bot, so allTriggers always only contains the trigger of the current bot